### PR TITLE
feat(onlyWhen): add onlyWhen

### PR DESCRIPTION
**Pull request by briancavalier**
Add `onlyWhen`

```hs
onlyWhen :: (a -> boolean) -> b -> (a -> b) -> a -> b
```

Given a predicate, default value, and function, return a function that evaluates the predicate and if true, applies the function, otherwise returns the default value.  Curried to 4 args so you can provide all arguments, including the input, if you want.

- [x] I added new tests for the issue I fixed or the feature I built
- [x] I ran `yarn test` for the package I'm modifying
- [ ] I used `yarn commit` instead of `git commit`. *Oops, I didn't*

### CODE REVIEW

File: src/function/onlyWhen/onlyWhen.test.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+describe(`onlyWhen`, () => {
+  describe(`given a predicate that is false`, () => {
+    it(`returns default value`, () => {
+      const expected = 0
+      assert.strictEqual(expected, onlyWhen((_: number) => false, expected, id, 1))
+    })
+  })
+
+  describe(`given a predicate that is false`, () => {
```
Comment (TylorS) [inline src/function/onlyWhen/onlyWhen.test.ts]: This test has the same name as above

File: src/function/onlyWhen/onlyWhen.test.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import * as assert from 'assert'
+
+import { id } from '..'
+import { onlyWhen } from './onlyWhen'
+
+describe(`onlyWhen`, () => {
+  describe(`given a predicate that is false`, () => {
+    it(`returns default value`, () => {
+      const expected = 0
+      assert.strictEqual(expected, onlyWhen((_: number) => false, expected, id, 1))
```
Comment (Frikki) [inline src/function/onlyWhen/onlyWhen.test.ts]: The order of arguments to `strictEqual()` is first `expected`, then `actual`, see https://nodejs.org/api/assert.html#assert_assert_strictequal_actual_expected_message.

Here they are in reverse order.

File: src/function/onlyWhen/onlyWhen.test.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    })
+  })
+
+  describe(`given a predicate that is true`, () => {
+    it(`returns f a`, () => {
+      const expected = 1
+      assert.strictEqual(2, onlyWhen((x: number) => x === expected, 0, (x: number) => x + 1, expected))
```
Comment (Frikki) [inline src/function/onlyWhen/onlyWhen.test.ts]: Same as above.

File: src/function/onlyWhen/onlyWhen.test.ts (hunk 2 above)
Comment (briancavalier) [inline src/function/onlyWhen/onlyWhen.test.ts]: ah, actual then expected.  I'd gotten used to putting expected first with https://github.com/briancavalier/assert.  Thanks, will fix.

**Comment by TylorS** [comment]
Thank you so much for the first outside contribution!! :heart: :rose:

**Pull request merged**